app/vep/vep_resources.py

```diff
@@ -88,3 +88,40 @@ async def vep_status(request: Request, submission_id: str):
     except Exception as e:
         logging.debug(e)
         return response_error_handler(result={"status": 500})
+
+@router.get("/download/{submission_id}", name="download_results")
+async def download_results(request: Request, submission_id: str):
+    try:
+        workflow_status_response = await get_workflow_status(submission_id)
+        workflow_status = workflow_status_response.dict()
+        # To use out file it will require changes in nextflow and get_workflow_status endpoint
+        results_file = workflow_status['outfile']
+        if workflow_status['status'] == "SUCCEDED":
+            return await FileResponse(results_file)
+        else:
+            response_msg = json.dumps(
+                {
+                    "status_code": 200,
```

Comment:
A 404 response with a status_code 200 in the body is evil. (This is why I don't understand the reason you include the status_code field in error responses.)

Reply:
This will only be sent as 200 response (on line 124). Errors are handled below.

Reply:
When I left my comment, the return value was:

<img width="571" alt="image" src="https://github.com/user-attachments/assets/271a3a17-2aa7-4981-b966-995411179118">

(see [commit d17c462](https://github.com/Ensembl/ensembl-web-tools-api/pull/26/commits/d17c4620a0429c0ec3efcfc28edea4e9104565ec#diff-d4096798e7cffcfe65307ad13fd12e1c101f522d30bd91a73fd58ffccae78086R108-R110))

Reply:
Also, reading through the code again, I do not think the successful response from the endpoint should include two different types of data (a file or a json). I think Kamal's initial version with a 404 response was correct.

Reply:
> I still don't understand why you guys are putting status codes in the response body though

Reply:
As the code currently stands, it can respond with a successful (200) response that can be either a file itself of a plain-text string saying that the file is not yet ready. This design worries me. It means that the client should be capable of inspecting the response and making decisions based on whether it's a file or a plain-text string.

Could we please change the design of the endpoint such that if the file is not available, the response is a 404? The message in the 404 response can say that the analysis for the submission with provided id has not yet finished.

Reply:
Fair point. This has been reflected now. I think download endpoint should return the file if the job is succeeded and else say 404 the result now found. The task to check the status of the workflow should lie with the /status endpoint and client should use that. otherwise download will become download + status both.
In this case client has requested a file if it exists then it should return it or else say file not found.